### PR TITLE
Added Glob::intersect()

### PR DESCRIPTION
**Pull request by EmanueleMinotto**
Should fix https://github.com/webmozart/glob/issues/7

### CODE REVIEW

File: src/Glob.php [hunk 1]
```diff
@@ -112,6 +112,84 @@ public static function glob($glob, $flags = 0)
...
+     * Combines two globs/paths into another, stricter glob/path.
+     *
+     * @param string $glob1 Path or glob expression.
+     * @param string $glob2 Path or glob expression.
+     *
+     * @return null|string
+     */
+    public static function combine($glob1, $glob2)
+    {
+        $d1 = static::toRegEx($glob1) === $glob1;
```
Comment (webmozart) [inline src/Glob.php]: You can use `isDynamic()` here, it's much more efficient

**Comment by webmozart** [comment]
Thanks a lot for this! :) I'll comment inline.

File: src/Glob.php [hunk 2]
```diff
@@ -112,6 +112,84 @@ public static function glob($glob, $flags = 0)
...
 
     /**
+     * Combines two globs/paths into another, stricter glob/path.
+     *
+     * @param string $exp1 Path or glob expression.
+     * @param string $exp2 Path or glob expression.
+     *
+     * @return null|string
+     */
+    public static function combine($exp1, $exp2)
```
Comment (webmozart) [inline src/Glob.php]: I think we should rename this to `intersect()`, that's easier to understand I guess. Also, please change `$exp1` and `$exp2` to `$glob1` and `$glob2`.

File: src/Glob.php [hunk 3]
```diff
@@ -112,6 +112,84 @@ public static function glob($glob, $flags = 0)
...
+    {
+        $s1 = !self::isDynamic($exp1);
```
Comment (webmozart) [inline src/Glob.php]: Please prefer expressive variable names: `$isStatic1`, `$isStatic2`

File: src/Glob.php [hunk 4]
```diff
@@ -112,6 +112,84 @@ public static function glob($glob, $flags = 0)
...
+        // return the path if it matches the glob and null otherwise.
+        if (!$s1 && $s2) {
+            return self::match($exp2, $exp1) ? $exp2 : null;
+        }
+        if ($s1 && !$s2) {
+            return self::match($exp1, $exp2) ? $exp1 : null;
+        }
+
+        // If both globs are dynamic, return the combined glob if they match each other and null otherwise.
+        if (!$s1 && !$s2) {
```
Comment (webmozart) [inline src/Glob.php]: You can invert the condition and return early here.

Comment (webmozart) [inline src/Glob.php]: Actually it should be possible to remove the if-statement entirely. It's always true, no?

**Comment by sebwebdev** [comment]
Nice one! A bit late to the implementation party, since I was busy working.
How about supporting '?' as wildcard, character classes and ranges?
See http://man7.org/linux/man-pages/man7/glob.7.html

**Comment by webmozart** [comment]
@sebwebdev Thanks for the suggestion! Could you open a new ticket? I think that's not really related to this ticket.

**Comment by webmozart** [comment]
Hi! :) Will you have time for working on more elaborate test cases? I'd like to merge this soon.

**Comment by EmanueleMinotto** [comment]
Oops... found some new tests, but sadly this seems to be a partial solution, I'm investigating on it.

**Comment by EmanueleMinotto** [comment]
@webmozart the solution is much more hard than what I did, look here: https://github.com/Pathgather/glob-intersection
I can work on it, but the solution won't arrive so soon :\

**Comment by webmozart** [comment]
Thanks for that link! Maybe this is too complicated. Do you want to work on a complete solution or should we drop this?

**Comment by EmanueleMinotto** [comment]
The big part of that repository is about the NFA management, time could be reduced adding a dependency like https://github.com/KendallHopkins/FormalTheory or something similar.
Yes it's complicated, but I can work on it :) the problem is that the solution won't be so fast

**Comment by webmozart** [comment]
To be honest I think that all this is out of the scope of this library. Probably it's even faster to get all files matching one glob first and then match their paths against the second glob (regex).

**Comment by EmanueleMinotto** [comment]
:+1:

**Pull request closed without merging**